### PR TITLE
Refactor method structure and definitions

### PR DESCRIPTION
**Pull request by jmazanec15**
### Description
This is a refactor around KNNMethod Definitions. It pre-work for interface for #1889. Functionally, nothing is changing. The overall scope/goal is to make it easier to integrate the quantization framework functionality and also make the plugin easier to maintain in the future. 

This refactor has 4 components (I did one commit per)
1. Remove getMethod from KNNLibrary interface. The motivation here is that the KNNMethod object is an implementation detail for an engine, so it does not make sense to expose it. The functionality exposed should be around validating configurations/params, and/or providing information relevant to building and/or searching the index produced by the KNNLibrary
2. Move EngineSpecificMethodContext into KNNMethod. Motivation is that the KNNMethod is meant to encapsulate a libraries ANN configuration. This includes the parameters necessary for search.
3. Change KNNMethod to interface and implement per Method we support. This resolves the dreaded configuration complexity we have in the engines by pulling the method definitions out into their own classes (see https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/index/engine/faiss/Faiss.java#L217-L311). 
4. Creating basic interface for Encoders. Similar to making KNNMethod an interface and implementing per method we support, I did the same with the Encoders. In the future, this will make it a lot easy to add Quantizers from the quantization framework. 

The next steps after this is to change the KNNLibrary.getMethodAsMap to return a new Object that will both contain information to be passed to the library as well as the map needed to build the index - specifically, it will return the information needed by the quantization framework to configure the quantization. This will simplify the experience for building more complex objects. 

### Related Issues
#1889 

### Check List
- [X] Commits are signed per the DCO using `--signoff`.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by navneet1v — DISMISSED** [review]
@jmazanec15 code looks good to me. At some point of time I really think we need a good class diagram on Method and Library class hierarchy for easy understanding. May be you can use IDE tool to build this and paste it with the PR. See if that is possible

**Comment by jmazanec15** [comment]
Sure Ill look into it - class hierarchy will help.

File: src/main/java/org/opensearch/knn/index/engine/faiss/FaissHNSWMethod.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    /**
+     * Constructor for FaissHNSWMethod
+     *
+     * @see AbstractKNNMethod
+     */
+    public FaissHNSWMethod() {
+        super(initMethodComponent(), Set.copyOf(SUPPORTED_SPACES), new DefaultHnswContext());
+    }
+
+    private static MethodComponent initMethodComponent() {
```
Comment (heemin32) [inline src/main/java/org/opensearch/knn/index/engine/faiss/FaissHNSWMethod.java:57]: nit. It doesn't have to be static.

Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/index/engine/faiss/FaissHNSWMethod.java:57]: its called for super constructor so it ends up actually having to

File: src/main/java/org/opensearch/knn/index/engine/Encoder.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+ * Interface representing an encoder. An encoder generally refers to a vector quantizer.
+ */
+public interface Encoder {
+    /**
+     * The name of the encoder does not have to be unique. Howevwer, when using within a method, there cannot be
+     * 2 encoders with the same name.
+     *
+     * @return Name of the encoder
+     */
+    String getName();
```
Comment (heemin32) [inline src/main/java/org/opensearch/knn/index/engine/Encoder.java:18]:
```suggestion
    String getName() {
      return getMEthodComponent().getName();
    }
```
and you can remove override function in all implementation class.

Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/index/engine/Encoder.java]: Makes sense. Added default

**Review by heemin32 — APPROVED** [review]
LGTM. Thanks!

**Review by navneet1v — APPROVED** [review]

**Review by naveentatikonda — APPROVED** [review]

**Pull request merged**